gnucash/import-export/import-main-matcher.cpp

```diff
@@ -192,32 +195,95 @@ defer_bal_computation (GNCImportMainMatcher *info, Account* acc)
+        gpointer rv;
+        gtk_tree_model_get(m_model, &*m_iter, column, &rv, -1);
+        return static_cast<T>(rv);
+    }
+    GNCImportTransInfo* get_trans_info ()
+    {
+        g_return_val_if_fail (m_iter.has_value(), nullptr);
+        return get<GNCImportTransInfo*>(DOWNLOADED_COL_DATA);
+    };
+    bool has_value() { return m_iter.has_value(); };
+    std::string get_account_name ()
+    {
+        g_return_val_if_fail (m_iter.has_value(), "");
+        auto str = get<char*>(DOWNLOADED_COL_ACCOUNT);
+        std::string rv{str};
+        g_free (str);
+        return rv;
+    }
+    GtkTreeModelIterator operator*() const { return *this; }
+    bool operator==(const GtkTreeModelIterator& other) const
+    {
+        if (!m_iter.has_value() && !other.m_iter.has_value())
+            return true;
+        if (!m_iter.has_value() || !other.m_iter.has_value())
+            return false;
+        return (m_model == other.m_model) &&
+               (m_iter->stamp == other.m_iter->stamp) &&
+               (m_iter->user_data == other.m_iter->user_data) &&
+               (m_iter->user_data2 == other.m_iter->user_data2) &&
+               (m_iter->user_data3 == other.m_iter->user_data3);
+    }
+    bool operator!=(const GtkTreeModelIterator& other) const { return !(*this == other); }
+private:
+    GtkTreeModel* m_model;
+    std::optional<GtkTreeIter> m_iter;
+};
+
+// Custom container class
+class GtkTreeModelContainer
+{
+public:
+    using iterator = GtkTreeModelIterator;
+    GtkTreeModelContainer(GtkTreeModel* model) : m_model(model), m_size{0}
+    {
+        for (auto it = begin(); it.has_value(); ++it)
+            m_size++;
+    }
+    iterator begin()
+    {
+        GtkTreeIter iter;
+        gtk_tree_model_get_iter_first(m_model, &iter);
+        return iterator(m_model, iter);
+    }
+    iterator end() {  return iterator(m_model, std::nullopt); }
+    size_t size() { return m_size; };
+    bool empty() { return m_size == 0; };
+private:
+    GtkTreeModel* m_model;
+    size_t m_size;
+};
+
 void
 gnc_gen_trans_list_delete (GNCImportMainMatcher *info)
 {
 
     if (info == NULL)
         return;
 
-    GtkTreeModel *model = gtk_tree_view_get_model (info->view);
-    GtkTreeIter iter;
-    if (gtk_tree_model_get_iter_first (model, &iter))
-    {
-        do
-        {
-            GNCImportTransInfo *trans_info;
-            gtk_tree_model_get (model, &iter,
-                                DOWNLOADED_COL_DATA, &trans_info,
-                                -1);
-
-            if (info->transaction_processed_cb)
-            {
-                info->transaction_processed_cb (trans_info, false,
-                                                info->user_data);
-            }
-        }
-        while (gtk_tree_model_iter_next (model, &iter));
-    }
+    if (info->transaction_processed_cb)
+        for (auto iter : GtkTreeModelContainer(gtk_tree_view_get_model (info->view)))
```

Comment:
instead of `for (auto iter : container)` I'd prefer `for (const auto& iter : container)` however the latter won't compile.

Reply:
That's because the result of `GtkTreeModelContainer(gtk_tree_view_get_model (info->view))` is a temporary and you're not allowed to take references to temporaries. There's apparently a lifetime adjustment in C++23 that will allow it in for loops, see Temporary range expression in [this](https://en.cppreference.com/w/cpp/language/range-for). It will be a few years before we can use that, so a work around would be to create a local and iterate on that:
```
auto info_view{GtkTreeModelContainer(gtk_tree_view_get_model (info->view))};
for (const auto& iter : info_view)
  //…
```

Reply:
Doesn't work...
```
/home/chris/sources/gnucash/stable/gnucash/import-export/import-main-matcher.cpp:309:27: error: cannot bind non-const lvalue reference of type 'ImportMatcherType&' to an rvalue of type 'ImportMatcherType'
  309 |         for (auto& iter : tree_model)
      |                           ^~~~~~~~~~
```

Reply:
Suggesting that you could bind a `const lvalue reference` but IIUC you already tried that. Looks like the compiler is optimizing away tree_model and leaving you with the temporary anyway. That probably means that `iter` is the temporary's memory and taking a reference wouldn't improve performance.